node/src/actors/chain_manager/handlers.rs

```diff
@@ -197,28 +189,41 @@ impl Handler<AddBlocks> for ChainManager {
         match self.sm_state {
             StateMachine::WaitingConsensus => {}
             StateMachine::Synchronizing => {
-                let old_chain_state = self.chain_state.clone();
                 if let Some(target_beacon) = self.target_beacon {
-                    for block in msg.blocks {
+                    let mut batch_succeded = true;
+                    for block in msg.blocks.iter() {
                         if let Err(e) = self.process_requested_block(ctx, block) {
-                            warn!("{}", e);
-                            self.chain_state = old_chain_state;
+                            log::error!("Error processing block: {}", e);
+                            self.initialize_from_storage(ctx);
+                            log::info!("Restored chain state from storage");
+                            batch_succeded = false;
+                            break;
+                        }
+
+                        if self.get_chain_beacon() == target_beacon {
                             break;
                         }
                     }
-                    let our_beacon = self
-                        .chain_state
-                        .chain_info
-                        .as_ref()
-                        .unwrap()
-                        .highest_block_checkpoint;
-                    if our_beacon == target_beacon {
+
+                    if batch_succeded {
+                        self.persist_chain_state(ctx);
+                        self.persist_blocks_batch(ctx, msg.blocks, target_beacon);
```

Comment:
These two lines produce subsequent synchronous writes to the storage without atomicity. I'd suggest reversing their order because I prefer my storage to temporally have orphan blocks rather than perpetually missing blocks that are actually referred in the block chain.